### ISSUE TITLE
training collapsed cifar10.3@40-1 with 4 GPUs

### ISSUE THREAD

**Issue by chychen** [google-research/fixmatch#10]
Dear authors,

any possible ideas result in following training collapsed?

docker environment: (TF 1.14.0) https://docs.nvidia.com/deeplearning/frameworks/tensorflow-release-notes/rel_19.10.html#rel_19.10

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 python fixmatch.py --filters=32 --dataset=cifar10.3@40-1 --train_dir ./experiments/fixmatch
```

![image](https://user-images.githubusercontent.com/5878561/73812394-989b8c80-4817-11ea-89e7-c51f957cad6f.png)

Thanks

**Comment by kihyuks** [comment]
Hello, thanks for reporting an issue. We did most of our developments using 1 GPU. We also didn't witness a training collapse before even with multi-GPU training, but need to mention that we didn't experiment as thoroughly with 4 GPUs.

In the meantime, we made some code changes to better support multi-GPU training. Would you please sync your branch and try re-running the code to see if the problem still exist?

**Comment by chychen** [comment]
@kihyuks problem solved! is that related to batchnorm?

![image](https://user-images.githubusercontent.com/5878561/74356224-f4a67800-4df8-11ea-8b08-4436ec26d0eb.png)
![image](https://user-images.githubusercontent.com/5878561/74356562-87471700-4df9-11ea-9c5a-05869c31d1b7.png)

### compare between 1gpu and 4gpu in this run:

In same train_step, 4gpu speedup by 2x than 1 gpu.
however, seems 1gpu converged slower but better.

**Comment by kihyuks** [comment]
Awesome! Yes, the recent fix is related to batch norm:

https://github.com/google-research/fixmatch/blob/bb377cb012543ce094cda114357184bc9b55a830/fixmatch.py#L99-L101
https://github.com/google-research/fixmatch/blob/bb377cb012543ce094cda114357184bc9b55a830/libml/utils.py#L225-L232

As mentioned in the paper (see Table 4), a variance induced by randomness of training is high when using 40 labeled training examples, so I would consider two results with 1 or 4 gpus are on a par with each other accuracy-wise.
